README.md
Library of ECDSA for BBc-1 transaction data
====
This library is originally from [BBc-1 repository](https://github.com/beyond-blockchain/bbc1), which is the library for signing/verifying BBc-1 transaction. 
When porting bbclib.py to another programming language, the dynamic link library generated in this project will be helpful.

# preparation (Linux/Mac)
```
sh prepare.sh
```
After finishing prepare.sh script, you will find libbbcsig.so (for Linux/Windows) or libbbcsig.dylib (for Mac) in lib/ directory.

# preparation (For AWS)
For Amazon Lambda/EC2, you need docker and do the following:
```
sh prepare.sh aws
```
After finishing prepare.sh script, you will find libbbcsig.so in the ami-docker/ directory.

If you want to generate static link library, do as follows:
```
cd lib
make libbbcsig.a
```
For example in golang, the static link library is useful to create stand alone exec command.


CHANGELOG.md
Change log
====

## v1.4
* implement new functions for importing private key and getting curve type parameter
  - This version will be used by py-bbclib and bbclib-go version 1.4.3 or later

## v1.2.3
* bug fix

## v1.2.2
* bug fix

## v1.2.1
* bug fix

## v1.2
* add static library

## v1.1.2
* bug fix

## v1.1.1
* memory free in sign() to avoid RSS wastage

## v1.1
* decoupled from BBc-1 main repository


tests/test_x509.py
import platform
import os
from ctypes import *
import binascii

if platform.system() == "Windows":
    lib = windll.LoadLibrary("../lib/libbbcsig.dll")
elif platform.system() == "Darwin":
    lib = cdll.LoadLibrary("../lib/libbbcsig.dylib")
else:
    lib = cdll.LoadLibrary("../lib/libbbcsig.so")

CURVETYPE = 2  # CURVE_TYPE_P256 in libbbcsig.h

POINT_CONVERSION_COMPRESSED = 2     # same as enum point_conversion_form_t in openssl/crypto/ec.h
POINT_CONVERSION_UNCOMPRESSED = 4   # same as enum point_conversion_form_t in openssl/crypto/ec.h

with open("private.key", "r") as f:
    private_key = f.read()

with open("self-signed.pem", "r") as f:
    pubkey_cert = f.read()

with open("private_expired.key", "r") as f:
    private_key2 = f.read()

with open("self-signed_expired.pem", "r") as f:
    pubkey_cert2 = f.read()


privkey_len = c_int32(32)
privkey     = (c_byte * privkey_len.value)()
pubkey_len  = c_int32(65)
pubkey      = (c_byte * pubkey_len.value)()

print("######### read private key only")
ret = lib.convert_from_pem(create_string_buffer(private_key.encode()), POINT_CONVERSION_UNCOMPRESSED, byref(pubkey_len), pubkey, byref(privkey_len), privkey)
print("private_key:", binascii.b2a_hex(privkey), ", len=", privkey_len.value)
print("public_key:", binascii.b2a_hex(pubkey), ", len=", pubkey_len.value)


privkey_len = c_int32(32)
privkey     = (c_byte * privkey_len.value)()
pubkey_len  = c_int32(65)
pubkey      = (c_byte * pubkey_len.value)()

print("\n######### read X509 public key cert")
ret = lib.read_x509(create_string_buffer(pubkey_cert.encode()), POINT_CONVERSION_UNCOMPRESSED, byref(pubkey_len), pubkey)
assert ret
print("public_key:", binascii.b2a_hex(pubkey), ", len=", pubkey_len.value)

print("\n######### verify X509 public key cert")
ret = lib.verify_x509(create_string_buffer(pubkey_cert.encode()), None)
assert ret == 1
print("verify ok")

print("\n######### verify X509 public key cert and private key")
ret = lib.verify_x509(create_string_buffer(pubkey_cert.encode()),
                      create_string_buffer(private_key.encode()))
assert ret == 1
print("verify ok")

print("\n######### verify X509 public key cert and different private key")
ret = lib.verify_x509(create_string_buffer(pubkey_cert.encode()),
                      create_string_buffer(private_key2.encode()))
assert ret == -4
print("invalid key pair (test ok)")


print("\n######### verify X509 public key cert (expired)")
ret = lib.verify_x509(create_string_buffer(pubkey_cert2.encode()), None)
assert ret == -5
print("expired")


tests/test_ecdsa.py
import platform
import binascii
import os
from ctypes import *

if platform.system() == "Windows":
    lib = windll.LoadLibrary("../lib/libbbcsig.dll")
elif platform.system() == "Darwin":
    lib = cdll.LoadLibrary("../lib/libbbcsig.dylib")
else:
    lib = cdll.LoadLibrary("../lib/libbbcsig.so")

CURVETYPE = 2  # CURVE_TYPE_P256 in libbbcsig.h

POINT_CONVERSION_COMPRESSED = 2     # same as enum point_conversion_form_t in openssl/crypto/ec.h
POINT_CONVERSION_UNCOMPRESSED = 4   # same as enum point_conversion_form_t in openssl/crypto/ec.h

test_digest = binascii.a2b_hex("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
privkey_len = c_int32(32)
privkey     = (c_byte * privkey_len.value)()
pubkey_len  = c_int32(65)
pubkey      = (c_byte * pubkey_len.value)()

signature = (c_byte * 64)()

print("# -- generate_keypair()")
lib.generate_keypair(CURVETYPE, POINT_CONVERSION_UNCOMPRESSED, byref(pubkey_len), pubkey, byref(privkey_len), privkey)
print("private_key:", binascii.b2a_hex(privkey), ", len=", privkey_len.value)
print("public_key:", binascii.b2a_hex(pubkey), ", len=", pubkey_len.value)

print("# -- sign()")
sig_r = (c_byte * 32)()
sig_s = (c_byte * 32)()
sig_r_len = (c_byte * 4)()  # Adjust size according to the expected size of sig_r and sig_s. Default:uint32.
sig_s_len = (c_byte * 4)()
ret = lib.sign(CURVETYPE, privkey_len, privkey, len(test_digest), test_digest, sig_r, sig_s, sig_r_len, sig_s_len)
print("sign:", ret)
assert ret == 1

sig_r_len = int.from_bytes(bytes(sig_r_len), "little")
sig_s_len = int.from_bytes(bytes(sig_s_len), "little")
sig_r = binascii.a2b_hex("00" * (32 - sig_r_len) + bytes(sig_r)[:sig_r_len].hex())
sig_s = binascii.a2b_hex("00" * (32 - sig_s_len) + bytes(sig_s)[:sig_s_len].hex())
signature = bytes(bytearray(sig_r)+bytearray(sig_s))
print(binascii.b2a_hex(signature))

print("# -- output_der()")
der_data = (c_byte * 512)()     # 256 -> 512
der_len = lib.output_der(CURVETYPE, privkey_len, privkey, byref(der_data))
print("DER: len=", der_len, "  dat=", binascii.b2a_hex(bytearray(der_data)[:der_len]))
assert der_len > 0

print("\n# -- clear private key")
privkey = (c_byte * privkey_len.value)()
print("private_key:", binascii.b2a_hex(privkey), ", len=", privkey_len.value)

print("# -- convert_from_der()")
pubkey_len = c_int32(33)
pubkey = (c_byte * pubkey_len.value)()
lib.convert_from_der(der_len, byref(der_data), POINT_CONVERSION_COMPRESSED, byref(pubkey_len), pubkey, byref(privkey_len), privkey)
print("private_key:", binascii.b2a_hex(privkey), ", len=", privkey_len.value)
print("public_key:", binascii.b2a_hex(pubkey), ", len=", pubkey_len.value)
assert privkey_len.value > 0
assert pubkey_len.value > 0

print("# -- sign()")
sig_r = (c_byte * 32)()
sig_s = (c_byte * 32)()
sig_r_len = (c_byte * 4)()  # Adjust size according to the expected size of sig_r and sig_s. Default:uint32.
sig_s_len = (c_byte * 4)()
ret = lib.sign(CURVETYPE, privkey_len, privkey, len(test_digest), test_digest, sig_r, sig_s, sig_r_len, sig_s_len)
print("sign:", ret)
assert ret == 1

sig_r_len = int.from_bytes(bytes(sig_r_len), "little")
sig_s_len = int.from_bytes(bytes(sig_s_len), "little")
sig_r = binascii.a2b_hex("00" * (32 - sig_r_len) + bytes(sig_r)[:sig_r_len].hex())
sig_s = binascii.a2b_hex("00" * (32 - sig_s_len) + bytes(sig_s)[:sig_s_len].hex())
signature = bytes(bytearray(sig_r)+bytearray(sig_s))
print(binascii.b2a_hex(signature))

print("\n# -- get compressed pubkey")
pubkey = (c_byte * 33)()
lib.get_public_key_compressed(CURVETYPE, privkey_len, privkey, byref(pubkey_len), pubkey)
print("public_key:", binascii.b2a_hex(pubkey), ", len=", pubkey_len.value)

print("# -- verify with compressed pubkey")
res = lib.verify(CURVETYPE, pubkey_len, pubkey, len(test_digest), test_digest, len(signature), signature)
print("result:", res)
assert res == 1

print("# -- output_public_key_der()")
der_data = (c_byte * 512)()     # 256 -> 512
der_len = lib.output_public_key_der(CURVETYPE, pubkey_len, pubkey, byref(der_data))
print("DER: len=", der_len, "  dat=", binascii.b2a_hex(bytearray(der_data)[:der_len]))
assert der_len > 0


print("\n# -- get uncompressed pubkey")
pubkey_len = c_int32(65)
pubkey = (c_byte * pubkey_len.value)()
lib.get_public_key_uncompressed(CURVETYPE, privkey_len, privkey, byref(pubkey_len), pubkey)
print("public_key:", binascii.b2a_hex(pubkey), ", len=", pubkey_len.value)

print("# -- verify with uncompressed pubkey")
res = lib.verify(CURVETYPE, pubkey_len, pubkey, len(test_digest), test_digest, len(signature), signature)
print("result:", res)
assert res == 1

print("# -- output_pem()")
pem = (c_char * 512)()      # 256 -> 512
length = lib.output_pem(CURVETYPE, privkey_len, privkey, byref(pem))
print("PEM: len=", length, " dat=", pem.value)
assert length > 0

print("# -- output_public_key_pem()")
pem = (c_char * 512)()      # 256 -> 512
length = lib.output_public_key_pem(CURVETYPE, pubkey_len, pubkey, byref(pem))
print("PEM: len=", length, " dat=", pem.value)
assert length > 0

print("# -- output_public_key_der()")
der_data = (c_byte * 512)()     # 256 -> 512
der_len = lib.output_public_key_der(CURVETYPE, pubkey_len, pubkey, byref(der_data))
print("DER: len=", der_len, "  dat=", binascii.b2a_hex(bytearray(der_data)[:der_len]))
assert der_len > 0

print("# -- output_der()")
der_data = (c_byte * 512)()     # 256 -> 512
der_len = lib.output_der(CURVETYPE, privkey_len, privkey, byref(der_data))
print("DER: len=", der_len, "  dat=", binascii.b2a_hex(bytearray(der_data)[:der_len]))
assert der_len > 0


print("******** successfully finished ********")


tests/create-test-keys.sh
openssl ecparam -out private.key -name prime256v1 -genkey
openssl req   -config openssl.cnf \
              -new \
              -x509 \
              -key private.key \
              -sha256 \
              -days 1 \
              -subj "/C=JP/ST=Tokyo/O=Beyond-Blockchain/CN=bbc1" \
              -out self-signed.pem


lib/libbbcsig.c
/**
  Copyright (c) 2017 beyond-blockchain.org.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

#include <stdio.h>
#include <stdbool.h>
#include <strings.h>

#include "libbbcsig.h"
#include <string.h>
#include <time.h>

#ifndef _WIN32
#include <strings.h>
#endif

#include <openssl/ec.h>      // for EC_GROUP_new_by_curve_name, EC_GROUP_free, EC_KEY_new, EC_KEY_set_group, EC_KEY_generate_key, EC_KEY_free
#include <openssl/ecdsa.h>   // for ECDSA_do_sign, ECDSA_do_verify
#include <openssl/bn.h>
#include <openssl/obj_mac.h> // for NID_secp192k1
#include <openssl/pem.h>
#include <openssl/evp.h>
#include <openssl/bio.h>
#include <openssl/err.h>
#include <openssl/x509.h>
#include <openssl/stack.h>
#include <openssl/x509_vfy.h>

#include <crypto/ec/ec_lcl.h>

#define CURVE_TYPE_SECP256  1
#define CURVE_TYPE_P256     2

#define macro_init_EC_GROUP(X)                                     \
    EC_GROUP *ecgroup;                                             \
    if (X == CURVE_TYPE_SECP256) {                                 \
        ecgroup = EC_GROUP_new_by_curve_name(NID_secp256k1);       \
    } else if (X == CURVE_TYPE_P256) {                             \
        ecgroup = EC_GROUP_new_by_curve_name(NID_X9_62_prime256v1);    \
    } else {                                                       \
        return false;          \
    }                          \
    if (NULL == ecgroup) {     \
        EC_KEY_free(eckey);    \
        return false;          \
    }                          \
    if (EC_KEY_set_group(eckey, ecgroup) != 1) { \
        EC_GROUP_free(ecgroup);                  \
        EC_KEY_free(eckey);                      \
        return false;                            \
    }

#define macro_init_EC_KEY(X)              \
    EC_KEY *eckey = EC_KEY_new();         \
    if (NULL == eckey) return false;      \
    macro_init_EC_GROUP(X)

#define macro_free_EC_KEY   \
    EC_GROUP_free(ecgroup); \
    EC_KEY_free(eckey)


static inline int _get_curvetype(int nid)
{
    if (nid == NID_secp256k1) return CURVE_TYPE_SECP256;
    if (nid == NID_X9_62_prime256v1) return CURVE_TYPE_P256;
    return 0;
}

static inline const BIGNUM * _get_naive_privateKey_from_eckey(EC_KEY *eckey, int *privkey_len, uint8_t *privkey)
{
    const BIGNUM *private_key = EC_KEY_get0_private_key(eckey);
    *privkey_len = BN_num_bytes(private_key);
    BN_bn2bin(private_key, privkey);
    return private_key;
}

static inline void _get_naive_pubicKey_from_eckey(EC_KEY *eckey, EC_GROUP *ecgroup, const uint8_t compression, int *pubkey_len, uint8_t *pubkey)
{
    BN_CTX *ctx = BN_CTX_new();
    const EC_POINT *pubkey_point = EC_KEY_get0_public_key(eckey);
    *pubkey_len = (compression == POINT_CONVERSION_UNCOMPRESSED) ? 65 : 33;
    EC_POINT_point2oct(ecgroup, pubkey_point, compression, pubkey, *pubkey_len, ctx);
    EC_POINT_free((EC_POINT *)pubkey_point);
    BN_CTX_free(ctx);
}

static inline void _calculate_publicKey_from_bignum_privateKey(EC_GROUP *ecgroup, const BIGNUM *private_key,
                                                               const uint8_t compression, int *pubkey_len, uint8_t *pubkey)
{
    EC_POINT *pubkey_point = EC_POINT_new(ecgroup);
    BN_CTX *ctx = BN_CTX_new();
    EC_POINT_mul(ecgroup, pubkey_point, private_key, NULL, NULL, ctx);

    *pubkey_len = (compression == POINT_CONVERSION_UNCOMPRESSED) ? 65 : 33;
    EC_POINT_point2oct(ecgroup, pubkey_point, compression, pubkey, *pubkey_len, ctx);

    BN_free((BIGNUM *)private_key);
    EC_POINT_free(pubkey_point);
    BN_CTX_free(ctx);
}

static inline void _build_eckey(EC_KEY *eckey, EC_GROUP *ecgroup, int privkey_len, uint8_t *privkey)
{
    BN_CTX *ctx = BN_CTX_new();
    BIGNUM *private_key = BN_bin2bn(privkey, privkey_len, NULL);
    EC_KEY_set_private_key(eckey, private_key);

    EC_POINT *pubkey_point = EC_POINT_new(ecgroup);
    EC_POINT_mul(ecgroup, pubkey_point, private_key, NULL, NULL, ctx);
    EC_KEY_set_public_key(eckey, pubkey_point);

    BN_free(private_key);
    EC_POINT_free(pubkey_point);
    BN_CTX_free(ctx);
}

static inline void _build_eckey_pub(EC_KEY *eckey, EC_GROUP *ecgroup, int point_len, uint8_t *point)
{
    BN_CTX *ctx = BN_CTX_new();
    EC_POINT *pubkey_point = EC_POINT_new(ecgroup);
    EC_POINT_oct2point(ecgroup, pubkey_point, point, point_len, ctx);
    EC_KEY_set_public_key(eckey, pubkey_point);
    EC_POINT_free(pubkey_point);
    BN_CTX_free(ctx);
}

static inline EVP_PKEY * _read_privateKey_pem(const char *pem)
{
    BIO* bo = BIO_new( BIO_s_mem() );
    BIO_write(bo, pem, strlen(pem));
    EVP_PKEY *privateKey = NULL;
    if (PEM_read_bio_PrivateKey( bo, &privateKey, 0, 0 ) == NULL) {
        privateKey = NULL;
    }
    BIO_free(bo);
    return privateKey;
}



VS_DLL_EXPORT
bool VS_STDCALL sign(const int curvetype, int privkey_len, uint8_t *privkey, int hash_len, uint8_t *hash,
                     uint8_t *sig_r, uint8_t *sig_t, uint32_t *sig_r_len, uint32_t *sig_s_len)
{
    macro_init_EC_KEY(curvetype);

    BIGNUM *private_key = BN_bin2bn(privkey, privkey_len, NULL);
    EC_KEY_set_private_key(eckey, private_key);
    BN_free(private_key);

    BN_CTX *ctx = BN_CTX_new();
    BIGNUM *kinv = NULL, *rp = NULL;
    bool flag = ECDSA_sign_setup(eckey, ctx, &kinv, &rp);
    BN_CTX_free(ctx);
    if (! flag) goto FIN;

    ECDSA_SIG *signature = ECDSA_do_sign_ex(hash, hash_len, kinv, rp, eckey);
    BN_bn2bin(signature->r, sig_r);
    BN_bn2bin(signature->s, sig_t);

    *sig_r_len = (uint32_t) BN_num_bytes(signature->r);
    *sig_s_len = (uint32_t) BN_num_bytes(signature->s);
    ECDSA_SIG_free(signature);

  FIN:
    BN_free(kinv);
    BN_free(rp);
    macro_free_EC_KEY;
    return flag;
}

VS_DLL_EXPORT
int VS_STDCALL verify(const int curvetype, int point_len, const uint8_t *point,
                      int hash_len, uint8_t *hash, int sig_len, const uint8_t *sig)
{
    macro_init_EC_KEY(curvetype);

    EC_POINT *pubkey_point = EC_POINT_new(ecgroup);
    BN_CTX *ctx = BN_CTX_new();
    EC_POINT_oct2point(ecgroup, pubkey_point, point, point_len, ctx);
    BN_CTX_free(ctx);

    EC_KEY_set_public_key(eckey, pubkey_point);
    EC_POINT_free(pubkey_point);

    ECDSA_SIG *signature = ECDSA_SIG_new();
    int numlen = (int)(sig_len/2);
    signature->r = BN_bin2bn(sig, numlen, NULL);
    signature->s = BN_bin2bn(&sig[numlen], numlen, NULL);

    int verify_status = ECDSA_do_verify(hash, hash_len, signature, eckey);

    ECDSA_SIG_free(signature);
    macro_free_EC_KEY;

    return verify_status;
}

VS_DLL_EXPORT
bool VS_STDCALL generate_keypair(const int curvetype, const uint8_t compression, int *pubkey_len, uint8_t *pubkey,
                      int *privkey_len, uint8_t *privkey)
{
    macro_init_EC_KEY(curvetype);

    EC_KEY_generate_key(eckey);
    const BIGNUM *private_key = _get_naive_privateKey_from_eckey(eckey, privkey_len, privkey);

    _get_naive_pubicKey_from_eckey(eckey, ecgroup, compression, pubkey_len, pubkey);

    BN_free((BIGNUM *)private_key);
    EC_GROUP_free(ecgroup);
    return true;
}

VS_DLL_EXPORT
bool VS_STDCALL get_public_key_uncompressed(const int curvetype, int privkey_len, uint8_t *privkey, int *pubkey_len, uint8_t *pubkey)
{
    macro_init_EC_KEY(curvetype);

    const BIGNUM *private_key = BN_bin2bn(privkey, privkey_len, NULL);
    _calculate_publicKey_from_bignum_privateKey(ecgroup, private_key, POINT_CONVERSION_UNCOMPRESSED, pubkey_len, pubkey);

    macro_free_EC_KEY;
    return true;
}

VS_DLL_EXPORT
bool VS_STDCALL get_public_key_compressed(const int curvetype, int privkey_len, uint8_t *privkey, int *pubkey_len, uint8_t *pubkey)
{
    macro_init_EC_KEY(curvetype);

    const BIGNUM *private_key = BN_bin2bn(privkey, privkey_len, NULL);
    _calculate_publicKey_from_bignum_privateKey(ecgroup, private_key, POINT_CONVERSION_COMPRESSED, pubkey_len, pubkey);

    macro_free_EC_KEY;
    return true;
}

VS_DLL_EXPORT
bool VS_STDCALL convert_from_der(long der_len, const unsigned char *der, const uint8_t compression,
                                 int *pubkey_len, uint8_t *pubkey, int *privkey_len, uint8_t *privkey)
{
    EC_KEY *eckey = d2i_ECPrivateKey(NULL, &der, der_len); // bug in memory allocation?
    if (eckey == NULL) return false;
    EC_GROUP *ecgroup = eckey->group;

    const BIGNUM *private_key = _get_naive_privateKey_from_eckey(eckey, privkey_len, privkey);

    _calculate_publicKey_from_bignum_privateKey(ecgroup, private_key, compression, pubkey_len, pubkey);

    EC_GROUP_free(ecgroup);
    return true;
}

VS_DLL_EXPORT
bool VS_STDCALL convert_from_der_with_curvetype(long der_len, const unsigned char *der, const uint8_t compression, int *curvetype,
                                                int *pubkey_len, uint8_t *pubkey, int *privkey_len, uint8_t *privkey)
{
    EC_KEY *eckey = d2i_ECPrivateKey(NULL, &der, der_len); // bug in memory allocation?
    if (eckey == NULL) return false;
    EC_GROUP *ecgroup = eckey->group;
    *curvetype = _get_curvetype(EC_GROUP_get_curve_name(ecgroup));

    const BIGNUM *private_key = _get_naive_privateKey_from_eckey(eckey, privkey_len, privkey);

    _calculate_publicKey_from_bignum_privateKey(ecgroup, private_key, compression, pubkey_len, pubkey);

    EC_GROUP_free(ecgroup);
    return true;
}

VS_DLL_EXPORT
bool VS_STDCALL convert_from_pem(const char *pem, const uint8_t compression,
                                 int *pubkey_len, uint8_t *pubkey, int *privkey_len, uint8_t *privkey)
{
    EVP_PKEY *privateKey =  _read_privateKey_pem(pem);

    EC_KEY *eckey = EVP_PKEY_get1_EC_KEY(privateKey);
    EC_GROUP *ecgroup = eckey->group;
    const BIGNUM *private_key = _get_naive_privateKey_from_eckey(eckey, privkey_len, privkey);

    _calculate_publicKey_from_bignum_privateKey(ecgroup, private_key, compression, pubkey_len, pubkey);

    macro_free_EC_KEY;
    return true;
}

VS_DLL_EXPORT
bool VS_STDCALL convert_from_pem_with_curvetype(const char *pem, const uint8_t compression, int *curvetype,
                                                int *pubkey_len, uint8_t *pubkey, int *privkey_len, uint8_t *privkey)
{
    EVP_PKEY *privateKey =  _read_privateKey_pem(pem);

    EC_KEY *eckey = EVP_PKEY_get1_EC_KEY(privateKey);
    EC_GROUP *ecgroup = eckey->group;
    *curvetype = _get_curvetype(EC_GROUP_get_curve_name(ecgroup));
    const BIGNUM *private_key = _get_naive_privateKey_from_eckey(eckey, privkey_len, privkey);

    _calculate_publicKey_from_bignum_privateKey(ecgroup, private_key, compression, pubkey_len, pubkey);

    macro_free_EC_KEY;
    return true;
}


static int cert_verify_callback(int ok, X509_STORE_CTX *ctx)
{
    /* Tolerate self-signed certificate */
    if (X509_STORE_CTX_get_error(ctx) == X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT) {
        return 1;
    }

    /* Otherwise don't override */
    return ok;
}


VS_DLL_EXPORT
bool VS_STDCALL read_x509(const char *pubkey_x509, const uint8_t compression,  int *pubkey_len, uint8_t *pubkey)
{
    BIO *bio = BIO_new(BIO_s_mem());
    BIO_write(bio, pubkey_x509, strlen(pubkey_x509));
    X509 *x509 = PEM_read_bio_X509(bio, NULL, NULL, NULL);
    if (x509 == NULL) {
        BIO_free(bio);
        return false;
    }
    BIO_free(bio);

    EVP_PKEY *public_key = X509_get_pubkey(x509);
    EC_KEY *eckey = EVP_PKEY_get1_EC_KEY(public_key);
    EC_GROUP *ecgroup = eckey->group;

    _get_naive_pubicKey_from_eckey(eckey, ecgroup, compression, pubkey_len, pubkey);

    EVP_PKEY_free(public_key);
    macro_free_EC_KEY;
    return true;
}

VS_DLL_EXPORT
bool VS_STDCALL read_x509_with_curvetype(const char *pubkey_x509, const uint8_t compression, int *curvetype, int *pubkey_len, uint8_t *pubkey)
{
    BIO *bio = BIO_new(BIO_s_mem());
    BIO_write(bio, pubkey_x509, strlen(pubkey_x509));
    X509 *x509 = PEM_read_bio_X509(bio, NULL, NULL, NULL);
    if (x509 == NULL) {
        BIO_free(bio);
        return false;
    }
    BIO_free(bio);

    EVP_PKEY *public_key = X509_get_pubkey(x509);
    EC_KEY *eckey = EVP_PKEY_get1_EC_KEY(public_key);
    EC_GROUP *ecgroup = eckey->group;
    *curvetype = _get_curvetype(EC_GROUP_get_curve_name(ecgroup));

    _get_naive_pubicKey_from_eckey(eckey, ecgroup, compression, pubkey_len, pubkey);

    EVP_PKEY_free(public_key);
    macro_free_EC_KEY;
    return true;
}

VS_DLL_EXPORT
int VS_STDCALL verify_x509(const char *pubkey_x509, const char *privkey_pem)
{
    int result;
    BIO *bio = BIO_new(BIO_s_mem());
    BIO_write(bio, pubkey_x509, strlen(pubkey_x509));
    X509 *x509 = PEM_read_bio_X509(bio, NULL, NULL, NULL);
    if (x509 == NULL) {
        BIO_free(bio);
        return EBADCERT;
    }

    X509_STORE *cert_store = NULL;
    X509_STORE_CTX *store_ctx = NULL;
    STACK_OF(X509) *stack_of_x509 = NULL;    // no need because of treating self-signed cert only
    time_t check_time;
    int store_ctx_error;
    int store_ctx_error_depth;

    if( (cert_store= X509_STORE_new()) == NULL) {
        result = EFAULURE;
        goto FIN;
    }
    if( (store_ctx= X509_STORE_CTX_new()) == NULL) {
        result = EFAULURE;
        X509_STORE_free(cert_store);
        goto FIN;
    }

    X509_STORE_set_verify_cb_func(cert_store, cert_verify_callback);
    if( !X509_STORE_CTX_init(store_ctx, cert_store, NULL, stack_of_x509) ) {
        result = EFAULURE;
        goto FIN2;
    }

    X509_STORE_CTX_set_cert(store_ctx, x509);
    time(&check_time);

    X509_STORE_CTX_set_time(store_ctx, 0, check_time);
    X509_STORE_CTX_set_flags(store_ctx, X509_V_FLAG_USE_CHECK_TIME);

    result = X509_verify_cert(store_ctx);
    if (result < 1) {
        result = EINVALIDCERT;
        store_ctx_error = X509_STORE_CTX_get_error(store_ctx);
        store_ctx_error_depth = X509_STORE_CTX_get_error_depth(store_ctx);
        //printf("Error %d at %d depth: %s\n", store_ctx_error, store_ctx_error_depth, X509_verify_cert_error_string(store_ctx_error));
        goto FIN2;
    }

    if (privkey_pem != NULL && strlen(privkey_pem) > 0) {
        EVP_PKEY *privateKey =  _read_privateKey_pem(privkey_pem);
        result = EBADKEYPAIR;
        if (privateKey) {
            result = X509_check_private_key(x509, privateKey);
            if (result == 0) result = EBADKEYPAIR;
        }
        EVP_PKEY_free(privateKey);
    }

  FIN2:
    X509_STORE_CTX_free(store_ctx);
    X509_STORE_free(cert_store);

  FIN:
    X509_free(x509);
    BIO_free(bio);

    if (result > 0) return 1;
    return result;
}


VS_DLL_EXPORT
int VS_STDCALL output_der(const int curvetype, int privkey_len, uint8_t *privkey, uint8_t *der_out)
{
    macro_init_EC_KEY(curvetype);

    _build_eckey(eckey, ecgroup, privkey_len, privkey);

    int der_len = i2d_ECPrivateKey(eckey, &der_out);

    macro_free_EC_KEY;
    return der_len;
}

VS_DLL_EXPORT
int VS_STDCALL output_pem(const int curvetype, int privkey_len, uint8_t *privkey, uint8_t *pem_out)
{
    macro_init_EC_KEY(curvetype);

    _build_eckey(eckey, ecgroup, privkey_len, privkey);

    BIO *out = BIO_new(BIO_s_mem());
    BUF_MEM *buf = BUF_MEM_new();
    memset(pem_out, 0, 512);
    PEM_write_bio_ECPrivateKey(out, eckey, NULL, NULL, 0, NULL, NULL);
    BIO_get_mem_ptr(out, &buf);

	int len = buf->length;
    memcpy(pem_out, buf->data, len);

    BIO_free_all(out);
    macro_free_EC_KEY;
    return len;
}

VS_DLL_EXPORT
int VS_STDCALL output_public_key_der(const int curvetype, int point_len, uint8_t *point, uint8_t *der_out)
{
    macro_init_EC_KEY(curvetype);

    _build_eckey_pub(eckey, ecgroup, point_len, point);

    int der_len = i2d_EC_PUBKEY(eckey, &der_out);

    macro_free_EC_KEY;
    return der_len;
}

VS_DLL_EXPORT
int VS_STDCALL output_public_key_pem(const int curvetype, int point_len, uint8_t *point, uint8_t *pem_out)
{
    macro_init_EC_KEY(curvetype);

    _build_eckey_pub(eckey, ecgroup, point_len, point);

    BIO *out = BIO_new(BIO_s_mem());
    BUF_MEM *buf = BUF_MEM_new();
    memset(pem_out, 0, 512);

    PEM_write_bio_EC_PUBKEY(out, eckey);
    BIO_get_mem_ptr(out, &buf);

	int len = buf->length;
    memcpy(pem_out, buf->data, len);

    BIO_free_all(out);
    macro_free_EC_KEY;
    return len;
}
